feat: add traefik plugin for request forwarding

PatrickMi opened the pull request:
Resolves #72

dargmuesli committed:
feat: stash

dargmuesli committed:
feat: stash

PatrickMi committed:
traefik changes

PatrickMi committed:
Add local plugin

PatrickMi committed:
Outcomment stomper and jobber

PatrickMi committed:
Adds local traefik auth plugin

Adds the local traefik auth plugin. The plugin connects to http://maevsi:3000/api/auth-proxy. But there are still problems with CORS and the turnstile headers

PatrickMi committed:
allow x-authenticate headers through cors check

PatrickMi committed:
feat(auth): add forward authentication plugin

PatrickMi committed:
Merge remote-tracking branch 'origin/master' into traefik_changes

dargmuesli commented on src/development/stack.yml:
```diff
@@ -333,6 +336,8 @@ services:
     - --providers.file.watch=true #DARGSTACK-REMOVE
     - --log.level=DEBUG #DARGSTACK-REMOVE
+    - --experimental.localPlugins.body-forward-auth.moduleName=github.com/PatrickMi/body-forward-auth
+    # - --experimental.plugins.body-forward-auth.version=0.1.1
```
Seems like Go is using an own versioning scheme with a prefixed `v` after all (https://go.dev/doc/modules/version-numbers). Maybe the plugin download will then work?
I recommended not using `v` as a prefix, because the automated releases for maevsi do not use it. Forgot about Go's custom scheme.

PatrickMi replied:
Thanks for that! It works now with the GitHub repository :-)

dargmuesli reviewed (changes requested):
Thank you for the submission!
What do you think of adding the CI pipeline and test from https://github.com/traefik/plugindemo?
Please remove the `.DS_Store` file from this PR.

PatrickMi committed:
Update src/development/plugins-local/src/github.com/PatrickMi/body-forward-auth/.traefik.yml

Co-authored-by: Jonas Thelemann <e-mail@jonas-thelemann.de>

PatrickMi committed:
feat(authentication): removes .DS_Store file

PatrickMi commented on src/development/plugins-local/src/github.com/PatrickMi/body-forward-auth/.traefik.yml:
```diff
@@ -0,0 +1,10 @@
+summary: 'Middleware Plugin that sends the request to another server so that it can perform authentication based on the headers and the body'
+
+testData:
+ AuthUrl : "127.0.0.1:3000"
```
Good point. I'll have a look to add the test to our pipeline

PatrickMi replied:
I'll change the indentation and add the `http://` as this was indeed missing.

PatrickMi commented:
> Thank you for the submission!
> What do you think of adding the CI pipeline and test from https://github.com/traefik/plugindemo?
> Please remove the `.DS_Store` file from this PR.

Good point. I'll have a look to add the test to our pipeline. 
I also deleted the `.DS_Store` file from the PR

PatrickMi committed:
feat(authentication): removes unused icon path

PatrickMi commented on src/development/plugins-local/src/github.com/PatrickMi/body-forward-auth/body_forward_auth.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+type BodyForwardAuth struct {
+	next    http.Handler
```
Good question. After first agreeing with you on the tab point I tried to setup the go formatter in VSCode and had some issues. After a short google search I stumbled upon this: https://go.dev/doc/effective_go#formatting. According to this document go has only a few rules regarding formatting. One of them is that go uses tabs instead of spaces for indentation. This formatting rule should be used if one wants to use the official go formatter as it is not a good style to change the formatting (as described here: https://github.com/microsoft/vscode-go/issues/2871#issuecomment-564833962). 

There is a package that can be installed to use a stricter go formatting ruleset (https://github.com/mvdan/gofumpt). However this also has no rule that enforces a blank line after a code block.

As I still agree with you on the point regarding the blank line after a block of code I added them by hand in the file.

PatrickMi committed:
feat(authentication): adds blank lines after code blocks

PatrickMi committed:
feat(authentication): remove unimportant logging

PatrickMi commented on src/development/plugins-local/src/github.com/PatrickMi/body-forward-auth/body_forward_auth.go:
```diff
@@ -0,0 +1,86 @@
+		http.Error(rw, err.Error(), http.StatusInternalServerError)
+		log.Printf("Error while reading the body: %e\n", err)
+	}
+	dump, err := httputil.DumpRequest(req, true)
```
I agree

PatrickMi committed:
feat(authentication): adds constant to enable/disable debug output

PatrickMi commented on src/development/plugins-local/src/github.com/PatrickMi/body-forward-auth/body_forward_auth.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func (bfa *BodyForwardAuth) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	log.Printf("The ServeHTTP function was called. Will read the body now.")
```
I agree. I added a constant to toggle debug logs

PatrickMi commented on src/development/plugins-local/src/github.com/PatrickMi/body-forward-auth/body_forward_auth.go:
```diff
@@ -0,0 +1,86 @@
+
+	if response.StatusCode >= 200 && response.StatusCode < 300 {
+		log.Printf("Got Response with status code: %d", response.StatusCode)
+		req.Body = io.NopCloser(bytes.NewBuffer(body))
```
This line creates a new ReadCloser object from the body object and inserts it back to the request. This is done because reading the body from the request in the first place removes the content from the reader. So it has to be added back to the body before the request is being forwarded to the database.

PatrickMi committed:
feat(authentication): removes unimportant logging

PatrickMi committed:
Update src/development/stack.env.template

Co-authored-by: Jonas Thelemann <e-mail@jonas-thelemann.de>

PatrickMi committed:
feat(authentication): removes todo

PatrickMi commented on src/production/secrets/maevsi/recaptcha-key.secret.template:
```diff
@@ -0,0 +1 @@
+<key>
```
This file is not used right now because a file containing the directory of a file containing the secret can't be given to the nuxt turnstile widget. I'll tweak the code and make a PR for this issue.

dargmuesli replied:
WIP: https://github.com/nuxt-modules/turnstile/pull/174

dargmuesli replied:
Turnstile v0.2.0 is released, the filename can now be changed :partying_face:

dargmuesli replied:
I implemented the file for turnstile

PatrickMi committed:
feat(authentication): moves plugin and rearrange code

moves the traefik plugin to the configurations/traefik folder,
moves the instantiation of the httpClient down to where it is used

PatrickMi commented on src/development/stack.yml:
```diff
@@ -234,14 +236,15 @@ services:
       - traefik.http.routers.postgraphile_secure.rule=Host(`postgraphile.${STACK_DOMAIN}`)
       - traefik.http.routers.postgraphile_secure.tls.options=mintls13@file #DARGSTACK-REMOVE
       - traefik.http.services.postgraphile.loadbalancer.server.port=5000
+      - traefik.http.middlewares.auth-plugin.plugin.body-forward-auth.AuthUrl=http://maevsi:3000/api/auth-proxy
```
I'll change it in a commit 👍

PatrickMi committed:
feat(authentication): rename plugin in stack configuration

PatrickMi committed:
feat(authentication): uses body-forward-auth plugin from github

dargmuesli commented on src/development/stack.yml:
```diff
@@ -365,6 +368,7 @@ services:
     - /var/run/docker.sock:/var/run/docker.sock
     - ./certificates/:/etc/traefik/acme/
     - ./configurations/traefik/dynamic.yml:/dynamic.yml:ro #DARGSTACK-REMOVE
+    # - ./configurations/traefik/plugins-local:/plugins-local
```
Can this line be removed?

PatrickMi replied:
Good find! I'll remove it as we don't use local traefik plugins anymore. 👍

dargmuesli commented on src/development/stack.env.template:
```diff
@@ -1,5 +1,6 @@
 GOOGLE_ANALYTICS_ID=
 PNPM_STORE_DIR=
+TURNSTILE_SITE_KEY=
```
Pnpm-Q-R-S-Turnstile :wink:

dargmuesli replied:
b216f7306dc0ab1bf25b1780c13a946ff60bb318

dargmuesli committed:
refactor(env): keep lines alphabetical

PatrickMi commented:
Resolves #72

dargmuesli committed:
Merge branch 'master' into traefik_changes

dargmuesli committed:
fix(turnstile): correct secret

PatrickMi committed:
chore(traefik): remove unnecessary comment

PatrickMi committed:
Merge remote-tracking branch 'origin/master' into traefik_changes

PatrickMi committed:
Merge branch 'master' into traefik_changes

dargmuesli commented on src/development/secrets/maevsi/turnstile-key.secret:
```diff
@@ -0,0 +1 @@
+1x0000000000000000000000000000000AA
```
Should this be changed to
```suggestion
1x0000000000000000000000000000000BB
```
as done in the [frontend](https://github.com/maevsi/maevsi/pull/1122/commits/ee9aa79cf920d297e6f199d295cbe8ee862a5002)?

dargmuesli replied:
As, I see.

PatrickMi committed:
fix(turnstile): use 'x-turnstile-key' as header for the ts key

dargmuesli reviewed (approved):
LGTM :rocket: can be merged when https://github.com/maevsi/maevsi/pull/1122 is ready!

dargmuesli committed:
fix(maevsi_beta): prepare for production
